Fix request interceptor to process all requests on same host as urlBase

partap opened the pull request:
Allow request interceptor to insert auth header on all requests matching the same host as urlBase.
The host includes the port, if supplied.
If no host is supplied for urlBase (e.g. `"/api"`) then location.host is used for the comparison.
All requests with no host automatically pass and get the auth header inserted.

This allows other endpoints of the loopback app (*not* in urlBase) to receive the AccessToken.

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

partap commented:
I couldn't get the tests to run on my installation... 
I tried:
`karma start karma.conf.js`
But it couldn't load framework:mocha or framework:junit

Do I need to do something else to run it?  Or can somebody check to make sure it passes?

partap commented:
Hrm...weird.  I noticed that there was a "test" script in package.json, so tried running it that way.  Still no dice:
```
partap@Donk:~/src/loopback-sdk-angular$ npm test

> loopback-sdk-angular@1.5.0 pretest /Users/partap/src/loopback-sdk-angular
> jshint . && bower install


> loopback-sdk-angular@1.5.0 test /Users/partap/src/loopback-sdk-angular
> node ./test.e2e/test-server.js node_modules/karma/bin/karma start --single-run --browsers PhantomJS

Test server is listening on http://localhost:3838/
Running node_modules/karma/bin/karma start --single-run --browsers PhantomJS
WARN [plugin]: Cannot find plugin "karma-phantomjs-reporter".
  Did you forget to install it ?
  npm install karma-phantomjs-reporter --save-dev
INFO [karma]: Karma v0.12.37 server started at http://localhost:9876/
WARN [launcher]: Can not load "PhantomJS", it is not registered!
  Perhaps you are missing some plugin?
```

partap commented:
Ok, the tests run now and it passes...

I had checked around on stackoverflow and added a `plugins` section to karma.conf.js and filled it with all the karma plugin names from node_modules...when I removed that section and ran `npm test` it worked.

partap commented:
@bajtos Does this solution work for you?  It solves *my* immediate problem, but it does bother me a bit that it is essentially hardcoded and non-configurable.

I was thinking back to your original solution of an explicit list of authorized hosts, as you suggested in (883e0512).

I think the simplest way would just be to add methods to `LoopBackResource`, similar to the existing `getUrlBase()` and `setUrlBase()`... maybe `setAuthHosts()` and `getAuthHosts()` to set and return the current array of hosts (and/or paths) that get the authorization header inserted. 

To address my earlier concern, I could normalize internal URLs in the interceptor using `{location.protocol}://{location.host}`.   Then we could just check each url for a prefix string match, much like the current mainline.  

Then in my case, I could set `urlBase` to `'/api/lb'` and `authHosts` to `['/']` ...I couldn't do that with the original suggestion because `'/'` prefix would have also also matched URLs of the form `'//externalsite.com/...'`

bajtos commented:
Sorry for the delay. This solution is fine with me, we can always refine it later, if there is demand.

I would like to see an automated unit-test for this though. One possible approach: add a new route to the test loopback app ([source](https://github.com/strongloop/loopback-sdk-angular/blob/b2b2d80c3e701a8fcaf5cb1a0a1d7c5480ecc452/test.e2e/test-server.js#L97-L98)), this route should return back info whether the request contained a valid access token or not. Or perhaps use `/api/users/me`, it should be available OOTB. Then you can call it via `$http.get`, using `localhost` and `127.0.0.1` - one request should fail with 401 and the second should pass.

Thoughts?

bajtos commented:
@slnode ok to test

partap commented:
I'm a little busy right now...I'll get some tests written when I have a chance... I'm thinking there should be a few more test cases though.  The existing one only tests the case of a request outside of the urlBase.

1. Request within `urlBase` *should* add accessToken header
2. Request not within `urlBase` but same host *should* add header
3. Request to different host *should not* add header

...and maybe something for internal requests, since the code treats them special (adding `"http://{window.host}"` to both `urlBase` and outgoing requests for comparision)

bajtos commented:
> I'm thinking there should be a few more test cases though. The existing one only tests the case of a request outside of the urlBase.

Cool! However, if some of the test cases you described above turn out to be difficult to setup, then I am fine with landing this PR without them. As long as there are at least *some* tests covering this change.

partap commented:
@bajtos OK, I left the existing test in place and added 3:

- intercepts internal requests with internal urlBase
- intercepts external requests to urlBase host
- does not intercept requests to non urlBase host

bajtos commented on test.e2e/spec/services.spec.js:
```diff
@@ -181,6 +181,89 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
+          var urlBase = '/api',
+              nonUrlBase = '/otherlocalpath',
+              authHeader = 'X-Test-Token',
+              accessTokenId = '12345';
```
`authHeader` and `accessTokenId` are duplicated in several tests. Could you please extract them to shared constants, e.g. `AUTH_HEADER` and `ACCESS_TOKEN_ID`

bajtos commented on test.e2e/spec/services.spec.js:
```diff
@@ -181,6 +181,89 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
+          var auth = $injector.get('LoopBackAuth');
+          auth.accessTokenId = accessTokenId;
+
+          var $http = $injector.get('$http');
```
L191-L200 are repeated in the test below too. Please extract them to a shared utility/helper method.

Example:

```js
function setupAuthenticatedHttpClient(urlBase, authHeader, tokenId) {
  loopBackResourceProvider.setUrlBase(urlBase);
  loopBackResourceProvider.setAuthHeader(authHeader);
  var $injector = createInjector();
  var auth = $injector.get('LoopBackAuth');
  auth.accessTokenId = accessTokenId;
  return $injector.get('$http');
}
```

Alternatively move (some of) these lines to a `beforeEach` hook. For example, I think it makes a lot of sense to call `setUrlBase` and `setAuthHeader` from a `beforeEach` hook.

bajtos commented on test.e2e/spec/services.spec.js:
```diff
@@ -181,6 +181,89 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
           );
         });
+
+        it('intercepts internal requests with internal urlBase',function() {
```
I find the terms `intercepts` and `does not intercept` difficult to understand when reading a test name. Perhaps `sends auth credentials` and `does not send auth credentials` would work better?

bajtos commented:
@partap thank you for the update. I left few comments where I thought the test code is not as clean/clear as it should be.

partap commented:
@bajtos Thanks, good suggestions.

Obviously, I just copypasta'd the original test a few times and tweaked them, so there were some typos and duplication.  

I didn't see the need to use a `beforeEach` hook since `urlBase` can vary per-test and we are calling the helper function for each anyway.

I went back and forth on whether to pass `AUTH_HEADER` and `ACCESS_TOKEN_ID` into`setupAuthenticatedHttpClient()` or to just use the constants.  I think works out cleaner using the constants because passing the values in each test bumps the line over the 80 column limit and it ends up adding either one or two lines per call, depending if your style guide calls for
```
setupAuthenticatedHttpClient(urlBase, AUTH_HEADER, 
                             ACCESS_TOKEN_ID);
```
or
```
setupAuthenticatedHttpClient(urlBase, 
                             AUTH_HEADER, 
                             ACCESS_TOKEN_ID);
```

...also, we decided the constants are constant for all the tests, right?  ;)

On the other hand, you could argue that the tests are clearer when you can see the constants passed into the function right before we start testing against those values... 

Your call.

partap commented:
Oh yeah...I added a test for the successful basic functionality of the interceptor: make sure it does insert credentials for a request *inside* the urlBase...

I left the original test alone, but it could probably be removed...I think the last test covers that case as well.

partap commented:
My bad... I thought I tested that before committing last night :(

partap commented:
Something wrong with the automatic tester?  It tests fine here.

bajtos commented:
@slnode test please

bajtos commented:
@partap LTGM now. Would you mind squashing the commits to a single one?

> Something wrong with the automatic tester? It tests fine here.

We are experiencing some issues with our Jenkins CI. I've triggered a new build, let's see how it will look.

bajtos commented on lib/services.template.ejs:
```diff
@@ -3,6 +3,15 @@
+function getHost(url) {
+  var m = url.match(/^(?:https?:)?\/\/([^\/]+)/);
+  if (m) {
+    return m[1]
```
Nitpick: missing `;`

bajtos commented on lib/services.template.ejs:
```diff
@@ -3,6 +3,15 @@
+  if (m) {
+    return m[1]
+  }
+  return null;
```
Actually, this can be probably simplified to

```js
return m ? m[1] : null;
```

Not a big deal though. You can either change this while rebasing, or ignore my comment (including the one above) and keep this as it is.

superkhau replied:
+1

superkhau commented on test.e2e/spec/services.spec.js:
```diff
@@ -155,32 +155,107 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
-              authHeader = 'X-Test-Token',
-              accessTokenId = '12345';
+        var AUTH_HEADER = 'X-Test-Token',
+            ACCESS_TOKEN_ID = '12345';
```
Nitpick: var per line

```
var AUTH_HEADER = 'X-Test-Token';
var ACCESS_TOKEN_ID = '12345';
```

superkhau commented on test.e2e/spec/services.spec.js:
```diff
@@ -155,32 +155,107 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
-          var $http = $injector.get('$http');
+        it('intercepts only configured urlBase requests', function() {
+          var urlBase = 'http://test.urlbase',
+              nonUrlBase = 'http://test.nonurlbase';
```
same here

superkhau commented on test.e2e/spec/services.spec.js:
```diff
@@ -155,32 +155,107 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
             function(req) {
               // when the request points to a non urlBase
-              // domain it should not have authHeader
-              expect(req.config.headers[authHeader]).to.equal(undefined);
```
to.be(undefined)

superkhau commented on test.e2e/spec/services.spec.js:
```diff
@@ -155,32 +155,107 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
-              // domain it should not have authHeader
-              expect(req.config.headers[authHeader]).to.equal(undefined);
+              // domain it should not have AUTH_HEADER
+              expect(req.config.headers[AUTH_HEADER]).to.equal(undefined);
```
to.be(undefined)

superkhau commented on test.e2e/spec/services.spec.js:
```diff
@@ -155,32 +155,107 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
+
+        it('sends auth credentials for urlBase requests', function() {
+          var urlBase = 'http://test.urlbase',
+              urlBaseResource = 'http://test.urlbase/resource';
```
same here

superkhau commented on test.e2e/spec/services.spec.js:
```diff
@@ -155,32 +155,107 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
+            function(req) {
+              // when the request points to a urlBase
+              // domain it should have AUTH_HEADER
+              expect(req.config.headers[AUTH_HEADER]).to.equal(ACCESS_TOKEN_ID);
```
to.be(undefined)

superkhau commented on test.e2e/spec/services.spec.js:
```diff
@@ -155,32 +155,107 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
+        it('sends auth creds for internal requests with internal urlBase',
+          function() {
+            var urlBase = '/api',
+                nonUrlBase = '/otherlocalpath';
```
same here

superkhau commented on test.e2e/spec/services.spec.js:
```diff
@@ -155,32 +155,107 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
+                  // and urlBase is an internal url
+                  // accessToken should be inserted
+                  expect(req.config.headers[AUTH_HEADER])
+                    .to.equal(ACCESS_TOKEN_ID);
```
to.be

superkhau replied:
nevermind this one is fine. ;)

superkhau commented on test.e2e/spec/services.spec.js:
```diff
@@ -155,32 +155,107 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
+        it('sends auth creds for external requests to urlBase host',
+          function() {
+            var urlBase = 'http://apihost/api',
+                nonUrlBase = 'http://apihost/otherpath';
```
same here

superkhau commented on test.e2e/spec/services.spec.js:
```diff
@@ -155,32 +155,107 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
+
+        it('does not send auth creds to non urlBase host', function() {
+          var urlBase = '/api',
+              nonUrlBaseHost = 'http://otherhost/api';
```
same here

superkhau commented on test.e2e/spec/services.spec.js:
```diff
@@ -155,32 +155,107 @@ define(['angular', 'given', 'util'], function(angular, given, util) {
+              function(req) {
+                // when the request points to a non urlBase
+                // domain it should not have AUTH_HEADER
+                expect(req.config.headers[AUTH_HEADER]).to.equal(undefined);
```
to.be

partap commented:
@superkhau I have no problem with those style changes (var-per-line, `.to.equal(undefined)` => `.to.be.undefined`)

But maybe they would be better handled in a separate PR, where you could change *all* such occurrences, rather than just in these new tests?

superkhau commented:
@partap These are just nitpicks anyways. I'll leave it up to @bajtos to decide what to do there. IMO though, since these are new tests, you might as well write them correctly the first time.

bajtos commented:
Yeah, I guess I should have been more careful when landing the existing tests. The new tests are copying the style present in one of the existing tests.

BTW I prefer `.to.equal(undefined)` over `.to.be.undefined` as property-getter-based assertions are evil.

bajtos commented:
I fixed variable declarations and improved the commit message to 1) follow [50/72 rule](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html) and 2) provide more info about the change.

Landed via 72529a2, thank you @partap for the contribution!

superkhau commented:
> TW I prefer .to.equal(undefined) over .to.be.undefined as property-getter-based assertions are evil.

Why are they evil?

partap commented:
>>BTW I prefer .to.equal(undefined) over .to.be.undefined as property-getter-based assertions are evil.

>Why are they evil?

@bajtos I'm curious about this as well...what did you mean, exactly?

My main objection would be using  `undefined` as a property name in the first place, but if the language allows it, and it makes logical sense (e.g. not overriding `undefined` to mean something different than `undefined`), I don't see the harm.
